edx_toggles/toggles/internal/waffle/flag.py

```diff
@@ -0,0 +1,179 @@
+    Represents a single waffle flag, enhanced with request-level caching.
+    """
+
+    _class_instances = WeakSet()
+
+    def __init__(self, name, module_name, log_prefix=""):
+        """
+        Waffle flag constructor
+        """
+        self.log_prefix = log_prefix
+        super().__init__(name, module_name)
+
+    def is_enabled(self):
+        """
+        Returns whether or not the flag is enabled.
+
+        Child classes that override this method should ensure that `set_monitor_value` is called for any
+        case not calling ``super().is_enabled()``.
+        """
+        value = self._get_flag_active()
+        self.set_monitor_value(value)
+        return value
+
+    @staticmethod
+    def cached_flags():
+        """
+        Returns a dictionary of all flags in the request cache. This method should only ever be used by child classes.
+        """
+        return _get_waffle_request_cache().setdefault("flags", {})
+
+    def _get_flag_active(self):
+        """
+        Return and cache the value of the flag activation. This does not handle monitoring.
+        """
+        # Check global cache
+        value = self.cached_flags().get(self.name)
+        if value is not None:
+            return value
+
+        # Check in context of request
+        request = crum.get_current_request()
+        value = self._get_flag_active_request(request)
+        if value is not None:
+            return value
+
+        # Return default value
+        return self._get_flag_active_no_request()
+
+    def _get_flag_active_request(self, request):
+        """
+        Get flag value in the context of the current request.
+        """
+        if request:
+            value = flag_is_active(request, self.name)
+            self.cached_flags()[self.name] = value
+            return value
+        return None
+
+    def _get_flag_active_no_request(self):
+        """
+        Return default value in the absence of any other, more specific flag value. This triggers warnings, as waffle
+        flag values are not supposed to be accessed in the absence of any request context.
+
+        Note: this skips the cache as the value might be different in a normal request context. This case seems to
+        occur when a page redirects to a 404, or for celery workers.
+        """
+        log.warning(
+            "%sFlag '%s' accessed without a request",
+            self.log_prefix,
+            self.name,
+        )
+        value = _is_flag_active_for_everyone(self.name)
+        set_custom_attribute("warn_flag_no_request_return_value", value)
+        return value
+
+    def set_monitor_value(self, value):
+        """
+        Send waffle flag value to monitoring. We keep this method such that it can be called by child classes (such as
+        edx-platform's waffle_utils.CourseWaffleFlag). Child classes are expected to call this method if they are
+        overriding `is_enabled`, and has cases where it is not calling `super().is_enabled().
```

Comment:
Second sentence is a mix of singular/plural that makes it a little hard to read.  I think switching `Child classes are...` to `A child class is..`, and fixing the rest to be singular, might read the best.